warehouse/query-core/src/main/java/datawave/query/iterator/logic/OrIterator.java

```diff
@@ -119,56 +141,68 @@ public T next() {
+                    if (!evaluationContext.equals(intersectExclude)) {
+                        candidateSet.add(evaluationContext);
+                    }
                 }
             }
-        }
 
-        // take the lowest of the candidates
-        if (candidateSet.size() > 0) {
-            lowest = candidateSet.first();
-
-            // decide how to construct the document
-            if (lowest.equals(lowestContextInclude)) {
-                // build it from the contextIncludeHeads
-                next = lowestContextInclude;
-                document = Util.buildNewDocument(contextIncludeHeads.get(next));
-            } else if (includeHeads.keySet().size() > 0 && lowest.equals(includeHeads.keySet().first())) {
-                // build it from the includeHeads
-                next = transforms.get(lowest);
-                document = Util.buildNewDocument(includeHeads.get(lowest));
-            } else {
-                // nothing to build it from all we know is that it wasn't in the exclude set
-                next = evaluationContext;
-                document = Util.buildNewDocument(Collections.emptyList());
-            }
+            // take the lowest of the candidates
+            if (candidateSet.size() > 0) {
+                lowest = candidateSet.first();
+                checkWaitWindow(lowest);
+
+                // decide how to construct the document
+                if (lowest.equals(lowestContextInclude)) {
+                    // build it from the contextIncludeHeads
+                    next = lowestContextInclude;
+                    document = Util.buildNewDocument(contextIncludeHeads.get(next));
+                } else if (includeHeads.keySet().size() > 0 && lowest.equals(includeHeads.keySet().first())) {
+                    // build it from the includeHeads
+                    next = transforms.get(lowest);
+                    document = Util.buildNewDocument(includeHeads.get(lowest));
+                } else {
+                    // nothing to build it from all we know is that it wasn't in the exclude set
+                    next = evaluationContext;
+                    document = Util.buildNewDocument(Collections.emptyList());
+                }
 
-            // regardless of where we hit make sure to advance includeHeads if it matches there
-            if (includeHeads != null && includeHeads.containsKey(lowest)) {
-                includeHeads = advanceIterators(lowest);
+                // regardless of where we hit make sure to advance includeHeads if it matches there
+                if (includeHeads != null && includeHeads.containsKey(lowest)) {
+                    includeHeads = advanceIterators(lowest);
+                }
+            }
+        } catch (WaitWindowOverrunException e) {
+            // if prev != null then it's a match that has not been returned
+            Key possibleYieldKey = null;
+            if (prev != null || !candidateSet.isEmpty()) {
```

Comment:
I think this needs to be `if (!isContextRequired() && (prev != null || !candidateSet.isEmpty()) {`

There is a case where the candidate set is populated but context is required to evaluate it `(A || !B)` or `(A || (!B && !C)`. In these cases the union can't actually map to a top key. An attempt to solve that problem is over in #2149

Reply:
But if prev != null, then that is a look-ahead key that was found to be a match and would have been returned next.  Assuming your logic is true....  shouldn't the change be:

`
if (prev != null || (!isContextRequired() && !candidateSet.isEmpty())
`

Reply:
Right you are!

Reply:
Committed change